pulsar-io/kinesis/src/main/java/org/apache/pulsar/io/kinesis/KinesisRecordProcessor.java

```diff
@@ -82,49 +93,66 @@ private void checkpoint(RecordProcessorCheckpointer checkpointer) {
         }
     }
 
+    public void updateSequenceNumberToCheckpoint(String sequenceNumber) {
+        this.sequenceNumberToCheckpoint = sequenceNumber;
+    }
+
+    public void failed() {
+        sourceContext.fatal(new PulsarClientException("Failed to process Kinesis records due send to pulsar topic"));
+    }
+
     @Override
     public void initialize(InitializationInput initializationInput) {
         kinesisShardId = initializationInput.shardId();
         log.info("Initializing KinesisRecordProcessor for shard {}. Config: checkpointInterval={}ms, numRetries={}, "
                         + "backoffTime={}ms, propertiesToInclude={}",
                 kinesisShardId, checkpointInterval, numRetries, backoffTime, propertiesToInclude);
+        nextCheckpointTimeInNanos = System.nanoTime() + checkpointInterval;
     }
 
     @Override
     public void processRecords(ProcessRecordsInput processRecordsInput) {
-
-        log.info("Processing " + processRecordsInput.records().size() + " records from " + kinesisShardId);
+        log.info("Processing {} records from {}", processRecordsInput.records().size(), kinesisShardId);
         long millisBehindLatest = processRecordsInput.millisBehindLatest();
 
         for (KinesisClientRecord record : processRecordsInput.records()) {
             try {
-                queue.put(new KinesisRecord(record, this.kinesisShardId, millisBehindLatest, propertiesToInclude));
+                queue.put(new KinesisRecord(record, this.kinesisShardId, millisBehindLatest,
+                        propertiesToInclude, this));
             } catch (InterruptedException e) {
                 log.warn("unable to create KinesisRecord ", e);
             }
         }
 
         // Checkpoint once every checkpoint interval.
         if (System.nanoTime() > nextCheckpointTimeInNanos) {
-            checkpoint(processRecordsInput.checkpointer());
+            if (sequenceNumberToCheckpoint != null
+                    && !sequenceNumberToCheckpoint.equals(lastCheckpointedSequenceNumber)) {
+                checkpoint(processRecordsInput.checkpointer(), sequenceNumberToCheckpoint);
+            }
```

Comment:
As we discussed privately, it's better to add a timer to invoke the checkpoint peridocally.